cpg_workflows/stages/realign_genotype_with_dragen.py

```diff
@@ -0,0 +1,329 @@
+    monitor_align_genotype_with_dragen,
+    prepare_ica_for_analysis,
+    run_align_genotype_with_dragen,
+)
+from cpg_workflows.targets import SequencingGroup
+from cpg_workflows.utils import slugify
+from cpg_workflows.workflow import SequencingGroupStage, StageInput, StageOutput, stage
+
+if TYPE_CHECKING:
+    from hailtop.batch.job import BashJob, PythonJob
+
+DRAGEN_VERSION: Final = config_retrieve(['ica', 'pipelines', 'dragen_version'])
+GCP_FOLDER_FOR_ICA_PREP: Final = f'ica/{DRAGEN_VERSION}/prepare'
+GCP_FOLDER_FOR_RUNNING_PIPELINE: Final = f'ica/{DRAGEN_VERSION}/pipelines'
+GCP_FOLDER_FOR_ICA_DOWNLOAD: Final = f'ica/{DRAGEN_VERSION}/output'
+ICA_REST_ENDPOINT: Final = 'https://ica.illumina.com/ica/rest'
+ICA_CLI_SETUP: Final = """
+mkdir -p $HOME/.icav2
+echo "server-url: ica.illumina.com" > /root/.icav2/config.yaml
+
+set +x
+gcloud secrets versions access latest --secret=illumina_cpg_workbench_api --project=cpg-common | jq -r .apiKey > key
+gcloud secrets versions access latest --secret=illumina_cpg_workbench_api --project=cpg-common | jq -r .projectID > projectID
+echo "x-api-key: $(cat key)" >> $HOME/.icav2/config.yaml
+icav2 projects enter $(cat projectID)
+set -x
+"""
+
+
+coloredlogs.install(level=logging.INFO)
+
+
+def calculate_needed_storage(
+    cram: str,
+) -> str:
+    logging.info(f'Checking blob size for {cram}')
+    storage_size: int = cpg_utils.to_path(cram).stat().st_size
+    return f'{ceil(ceil((storage_size / (1024**3)) + 3) * 1.2)}Gi'
+
+
+# No need to register this stage in Metamist I think, just ICA prep
+@stage
+class PrepareIcaForDragenAnalysis(SequencingGroupStage):
+    """Set up ICA for a single realignment run.
+
+    Creates a folder ID for the Dragen output to be written into.
+    """
+
+    def expected_outputs(self, sequencing_group: SequencingGroup) -> cpg_utils.Path:
+        sg_bucket: cpg_utils.Path = sequencing_group.dataset.prefix()
+        return sg_bucket / GCP_FOLDER_FOR_ICA_PREP / f'{sequencing_group.name}_output_fid.json'
+
+    def queue_jobs(self, sequencing_group: SequencingGroup, inputs: StageInput) -> StageOutput | None:
+        bucket_name: str = get_path_components_from_gcp_path(path=str(object=sequencing_group.cram))['bucket']
+
+        prepare_ica_job: PythonJob = get_batch().new_python_job(
+            name='PrepareIcaForDragenAnalysis',
+            attributes=(self.get_job_attrs() or {}) | {'tool': 'ICA'},
+        )
+        prepare_ica_job.image(image=image_path('cpg_workflows'))
+
+        outputs = self.expected_outputs(sequencing_group=sequencing_group)
+        output_fids = prepare_ica_job.call(
+            prepare_ica_for_analysis.run,
+            ica_analysis_output_folder=config_retrieve(['ica', 'data_prep', 'output_folder']),
+            api_root=ICA_REST_ENDPOINT,
+            sg_name=sequencing_group.name,
+            bucket_name=bucket_name,
+        ).as_json()
+
+        get_batch().write_output(
+            output_fids,
```

Comment:
just out of curiosity is this result being kept as a `{key: result_ID}` dictionary because it needs to be passed that way in future? If the json'ification isn't important I would have assumed just returning/storing the ID in isolation would be easier (not needing JQ/other to read back later)

Reply:
Yeah that'd be a better solution. Earlier this method had multiple key: value pairs, so json did make sense, but that changed once I had to redo the upload method. Will change

Reply:
I will look into changing this (as that would be better) in a new branch, because it'd be good to get something working merged in first.